source/server/listener_manager_impl.cc

```diff
@@ -619,41 +624,57 @@ std::vector<std::reference_wrapper<Network::ListenerConfig>> ListenerManagerImpl
-        //                     probably want this at Lyft but I will do it in a follow up.
-        ENVOY_LOG(critical, "listener '{}' failed to listen on address '{}' on worker",
-                  listener.name(), listener.listenSocketFactory().localAddress()->asString());
-        stats_.listener_create_failure_.inc();
-        removeListener(listener.name());
-      }
-      if (success) {
+  if (overridden_listener.has_value()) {
+    ENVOY_LOG(debug, "replacing existing listener {}", overridden_listener.value());
+    worker.addListener(overridden_listener, listener, [this, completion_callback](bool) -> void {
+      server_.dispatcher().post([this, completion_callback]() -> void {
         stats_.listener_create_success_.inc();
-      }
-      if (completion_callback) {
-        completion_callback();
-      }
+        if (completion_callback) {
+          completion_callback();
+        }
+      });
     });
-  });
+    return;
+  }
+  worker.addListener(
+      overridden_listener, listener, [this, &listener, completion_callback](bool success) -> void {
+        // The add listener completion runs on the worker thread. Post back to the main thread to
+        // avoid locking.
+        server_.dispatcher().post([this, success, &listener, completion_callback]() -> void {
+          // It is theoretically possible for a listener to get added on 1 worker but not the
+          // others. The below check with onListenerCreateFailure() is there to ensure we execute
+          // the removal/logging/stats at most once on failure. Note also that drain/removal can
+          // race with addition. It's guaranteed that workers process remove after add so this
+          // should be fine.
+          if (!success && !listener.onListenerCreateFailure()) {
+            // TODO(mattklein123): In addition to a critical log and a stat, we should consider
+            // adding a startup option here to cause the server to exit. I think we probably want
+            // this at Lyft but I will do it in a follow up.
+            ENVOY_LOG(critical, "listener '{}' failed to listen on address '{}' on worker",
+                      listener.name(), listener.listenSocketFactory().localAddress()->asString());
+            stats_.listener_create_failure_.inc();
+            removeListener(listener.name());
+          }
+          if (success) {
+            stats_.listener_create_success_.inc();
+          }
+          if (completion_callback) {
+            completion_callback();
+          }
+        });
+      });
 }
 
 void ListenerManagerImpl::onListenerWarmed(ListenerImpl& listener) {
   // The warmed listener should be added first so that the worker will accept new connections
   // when it stops listening on the old listener.
   for (const auto& worker : workers_) {
-    addListenerToWorker(*worker, listener, /* callback */ nullptr);
+    addListenerToWorker(*worker, /*overridden_listener=*/absl::nullopt, listener,
```

Comment:
nit: in general we don't do this (parameter name comments), please delete here and elsewhere.